src/Expander/Primitives.hs

```diff
@@ -0,0 +1,578 @@
+  valStxDest <- schedule (Ty TSyntax) valStx
+  linkExpr dest $ CoreReplaceLoc locDest valStxDest
+
+syntaxCase :: ExprPrim
+syntaxCase t dest stx = do
+  Stx scs loc (_ :: Syntax, args) <- mustBeCons stx
+  Stx _ _ (scrutinee, patterns) <- mustBeCons (Syntax (Stx scs loc (List args)))
+  scrutDest <- schedule (Ty TSyntax) scrutinee
+  patternDests <- traverse (mustHaveEntries >=> expandPatternCase t) patterns
+  linkExpr dest $ CoreCase scrutDest patternDests
+
+letSyntax :: ExprPrim
+letSyntax t dest stx = do
+  Stx _ loc (_ :: Syntax, macro, body) <- mustHaveEntries stx
+  Stx _ _ (mName, mdef) <- mustHaveEntries macro
+  sc <- freshScope $ T.pack $ "Scope for let-syntax at " ++ shortShow loc
+  m <- mustBeIdent mName
+  p <- currentPhase
+  -- Here, the binding occurrence of the macro gets the
+  -- fresh scope at all phases, but later, the scope is only
+  -- added to the correct phase in potential use sites.
+  -- This prevents the body of the macro (in an earlier
+  -- phase) from being able to refer to the macro itself.
+  let m' = addScope' m sc
+  b <- freshBinding
+  addLocalBinding m' b
+  v <- freshMacroVar
+  macroDest <- inEarlierPhase $ do
+    psc <- phaseRoot
+    schedule (Ty (TFun (Ty TSyntax) (Ty (TMacro (Ty TSyntax)))))
+      (addScope (prior p) mdef psc)
+  forkAwaitingMacro b v m' macroDest (ExprDest t dest) (addScope p body sc)
+
+log :: ExprPrim
+log t dest stx = do
+  unify dest (Ty (TMacro (Ty TUnit))) t
+  Stx _ _ (_ :: Syntax, message) <- mustHaveEntries stx
+  msgDest <- schedule (Ty TSyntax) message
+  linkExpr dest $ CoreLog msgDest
+
+dataCase :: ExprPrim
+dataCase t dest stx = do
+  Stx _ _ (_, scrut, cases) <- mustBeConsCons stx
+  a <- Ty . TMetaVar <$> freshMeta
+  scrutineeDest <- schedule a scrut
+  cases' <- traverse (mustHaveEntries >=> scheduleDataPattern t a) cases
+  linkExpr dest $ CoreDataCase scrutineeDest cases'
+
+---------------------
+-- Type Primitives --
+---------------------
+type TypePrim =
+  SplitTypePtr -> Syntax -> Expand ()
+
+baseType :: TyF SplitTypePtr -> TypePrim
+baseType ctor dest stx = do
+  _actualName <- mustBeIdent stx
+  linkType dest ctor
+
+arrowType :: TypePrim
+arrowType dest stx = do
+  Stx _ _ (_ :: Syntax, arg, ret) <- mustHaveEntries stx
+  argDest <- scheduleType arg
+  retDest <- scheduleType ret
+  linkType dest (TFun argDest retDest)
+
+macroType :: TypePrim
+macroType dest stx = do
+  Stx _ _ (_ :: Syntax, t) <- mustHaveEntries stx
+  tDest <- scheduleType t
+  linkType dest (TMacro tDest)
+
+-------------
+-- Modules --
+-------------
+
+makeModule :: DeclExpander -> Syntax -> Expand ()
+makeModule expandDeclForms stx =
+  view expanderModuleTop <$> getState >>=
+  \case
+    Just _ ->
+      error "TODO throw real error - already expanding a module"
+    Nothing -> do
+      bodyPtr <- newDeclTreePtr
+      modifyState $ set expanderModuleTop (Just bodyPtr)
+      Stx _ _ (_ :: Syntax, name, body) <- mustHaveEntries stx
```

Comment:
This line used to ask for `imports` and `exports` in addition to `name` and `body`. `resolveImports imports` and `buildExports exports` were both no-ops, so removing them makes sense, but I'm surprised not to see any corresponding change to any of the examples. Does that mean that the `#%module` primitive is untested?

Reply:
I created an issue, #40.